src/Build/BackEnd/BuildManager/GraphBuildSubmission.cs

```diff
@@ -0,0 +1,161 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Threading;
+using Microsoft.Build.Shared;
+using System.Globalization;
+using Microsoft.Build.Exceptions;
+using Microsoft.Build.Graph;
+
+namespace Microsoft.Build.Execution
+{
+    /// <summary>
+    /// A callback used to receive notification that a build has completed.
+    /// </summary>
+    /// <remarks>
+    /// When this delegate is invoked, the WaitHandle on the BuildSubmission will have been be signalled and the OverallBuildResult will be valid.
+    /// </remarks>
+    public delegate void GraphBuildSubmissionCompleteCallback(GraphBuildSubmission submission);
+
+    /// <summary>
+    /// A GraphBuildSubmission represents a graph build request which has been submitted to the BuildManager for processing.  It may be used to
+    /// execute synchronous or asynchronous graph build requests and provides access to the results upon completion.
+    /// </summary>
+    /// <remarks>
+    /// This class is thread-safe.
+    /// </remarks>
+    public class GraphBuildSubmission
+    {
+        /// <summary>
+        /// The callback to invoke when the submission is complete.
+        /// </summary>
+        private GraphBuildSubmissionCompleteCallback _completionCallback;
+
+        /// <summary>
+        /// The completion event.
+        /// </summary>
+        private readonly ManualResetEvent _completionEvent;
+
+        /// <summary>
+        /// The build results for each node.
+        /// </summary>
+        /// <remarks>
+        /// This collection may not contain results for all nodes until the submission is complete.
+        /// </remarks>
+        private readonly ConcurrentDictionary<ProjectGraphNode, BuildResult> _results = new ConcurrentDictionary<ProjectGraphNode, BuildResult>();
```

Comment:
Could this be a GraphBuildResult instead?

Reply:
Lol, this field is never used. Probably a result of some refactoring and I left this hanging.

Reply:
Your other PR looks over sadly, shaking its head.